src/mgds/pipelineModules/DropTags.py

```diff
@@ -0,0 +1,155 @@
+            keep_tags_count_in_name: str,
+            text_out_name: str,
+    ):
+        super(DropTags, self).__init__()
+        self.text_in_name = text_in_name
+        self.enabled_in_name = enabled_in_name
+        self.probability_in_name = probability_in_name
+        self.dropout_mode_in_name = dropout_mode_in_name
+        self.special_tags_in_name = special_tags_in_name
+        self.special_tag_mode_in_name = special_tag_mode_in_name
+        self.delimiter_in_name = delimiter_in_name
+        self.regex_enabled_in_name = regex_enabled_in_name
+        self.keep_tags_count_in_name = keep_tags_count_in_name
+        self.text_out_name = text_out_name
+
+    def length(self) -> int:
+        return self._get_previous_length(self.text_in_name)
+
+    def get_inputs(self) -> list[str]:
+        return [self.text_in_name, self.enabled_in_name, self.probability_in_name, self.dropout_mode_in_name, self.special_tags_in_name, self.special_tag_mode_in_name, self.delimiter_in_name, self.keep_tags_count_in_name]
+
+    def get_outputs(self) -> list[str]:
+        return [self.text_out_name]
+    
+    #convert special_tags to list depending on whether it's a newline-separated csv/txt file or a delimiter-separated string
+    #cached to reduce file read operations
+    @functools.lru_cache
+    def get_special_tags(self, sptags, delim):
+        if (sptags.endswith(".txt") and os.path.isfile(sptags)):
+            with open(sptags) as special_tags_file:
+                return [line.rstrip('\n') for line in special_tags_file]
+        elif (sptags.endswith(".csv") and os.path.isfile(sptags)):
+            with open(sptags, 'r') as special_tags_file:
+                splist = []
+                for row in csv.reader(special_tags_file):
+                    splist.append(row[0])
+                return splist
+        else:
+            return [tag.strip() for tag in sptags.split(delim)]
+
+    #parse regex expressions, create new special list based on matches   
+    def parse_regex(self, splist_in, taglist):
+        splist_out = []
+        regex_spchars = set(".^$*+?!\{\}\[\]|()\\")
+        for c in splist_in:
+            if any((a in regex_spchars) for a in c):    #only do regex matching if tag contains special character
+                c = c.replace("\)", "\\\\\)")
+                c = c.replace("\(", "\\\\\(")   #hopefully fix issues caused by "\(\)" syntax without affecting other regex
+                r = re.compile(c)
+                for s in taglist:
+                    if r.fullmatch(s):
+                        splist_out.append(s)
+            else:
+                splist_out.append(c)
+        return splist_out    
+    
+    #change probability evaluated against random() depending on mode
+    def probability_weighted(self, p, mode, i, len):
+        if mode == "RANDOM":
+            return float(p)
+        elif mode == "RANDOM WEIGHTED":
+            return float(p*(i/len))
```

Comment:
What does "weighted" mean here? weighted by what?

Reply:
The "weighting" in that mode is referring to the probability of a tag being dropped being weighted by its position in the full caption, with tags near the front being less likely to be dropped and tags near the end being more likely. It takes the fixed "probability" value and multiplies it by "i" (position in the list) divided by the length of the list, effectively reducing the probability at the start of the list and only reaching the full value at the end. For example, with a caption "a, b, c, d, e, f" with keep_tags=1 and probability 0.5, "RANDOM" would never drop "a" but would have a 50% chance to drop "b", "c", etc. "RANDOM WEIGHTED" would also never drop "a", but it would only have a 10% chance to drop "b", 20% for "c", 30% for "d", 40% for "e", and 50% for "f".

I included this mode as a way to allow more "important" tags at the front to be kept more often, with less important tags at the end to be dropped more. Most autotaggers will organize tags by descending confidence, which isn't necessarily the same thing as "importance" but may have a similar benefit for dropping. Or the user can manually order tags by their own metric of importance, based on categories or some subjective criteria. "Weighted" may not be the best description of that behavior, so if you have a suggestion for a different term I'm open to changing it both here and on the Onetrainer side.